lib/ngx/re.lua

```diff
@@ -0,0 +1,154 @@
+-- I hereby assign copyright in this code to the lua-resty-core project,
+-- to be licensed under the same terms as the rest of the code.
+
+local ffi = require 'ffi'
+local bit = require "bit"
+local base = require "resty.core.base"
+local core_regex = require "resty.core.regex"
+
+
+local C = ffi.C
+local sub = string.sub
+local band = bit.band
+local new_tab = base.new_tab
+local tostring = tostring
+local re_match_compile = core_regex.re_match_compile
+local destroy_compiled_regex = core_regex.destroy_compiled_regex
+
+
+local PCRE_ERROR_NOMATCH = -1
+
+
+local FLAG_DFA           = 0x02
+
+
+ffi.cdef [[
+    int ngx_http_lua_ffi_exec_regex(ngx_http_lua_regex_t *re, int flags,
+        const unsigned char *s, size_t len, int pos);
+]]
+
+
+local _M = { version = base.version }
+
+
+local function re_split_helper(subj, regex, opts, ctx)
+    local compiled, compile_once, flags = re_match_compile(regex, opts)
+    if compiled == nil then
+        -- compiled_once holds the error string
+        return nil, nil, nil, compile_once
+    end
+
+    -- exec the compiled regex
+
+    local rc
+    do
+        local pos = ctx.pos - 1
+        if pos < 0 then
+            pos = 0
+        end
+
+        rc = C.ngx_http_lua_ffi_exec_regex(compiled, flags, subj, #subj, pos)
+    end
+
+    if rc == PCRE_ERROR_NOMATCH then
+        if not compile_once then
+            destroy_compiled_regex(compiled)
+        end
+        return nil, nil, nil
+    end
+
+    if rc < 0 then
+        if not compile_once then
+            destroy_compiled_regex(compiled)
+        end
+        return nil, nil, nil, "pcre_exec() failed: " .. rc
+    end
+
+    if rc == 0 then
+        if band(flags, FLAG_DFA) == 0 then
+            return nil, nil, nil, "capture size too small"
+        end
+
+        rc = 1
+    end
+
+    local caps = compiled.captures
+    local ncaps = compiled.ncaptures
+
+    local from = caps[0] + 1
+    local to = caps[1]
+
+    if from < 0 or to < 0 then
+        return nil, nil, nil
+    end
+
+    ctx.pos = to + 1
+
+    -- retrieve the first sub-match capture if any
+
+    if ncaps > 0 and rc > 1 then
+        return from, to, sub(subj, caps[2] + 1, caps[3])
+    end
+
+    return from, to
+end
+
+
+function _M.split(subj, regex, opts, ctx, max, res)
+    -- we need to cast this to strings to avoid exceptions when they are
+    -- something else.
+    -- needed because of further calls to string.sub in this function.
+    subj = tostring(subj)
+
+    if not ctx then
+        ctx = new_tab(0, 1)
```

Comment:
I think instead of creating our own `ctx` table upon every `ngx.re.split` call without a user `ctx` argument (which is common) here, we should share the `ctx` table on the upper level. This is safe since this `split` function never ever yield so there is no possibility for conflicts.